core/src/test/java/ai/timefold/solver/core/preview/api/variable/declarative/fsr/FSRShadowVariableTest.java

```diff
@@ -0,0 +1,491 @@
+package ai.timefold.solver.core.preview.api.variable.declarative.fsr;
+
+import static ai.timefold.solver.core.impl.testdata.domain.declarative.fsr.TestdataFSRVisit.BASE_START_TIME;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verifyNoInteractions;
+
+import java.util.List;
+
+import ai.timefold.solver.core.api.solver.SolverFactory;
+import ai.timefold.solver.core.config.score.director.ScoreDirectorFactoryConfig;
+import ai.timefold.solver.core.config.solver.EnvironmentMode;
+import ai.timefold.solver.core.config.solver.PreviewFeature;
+import ai.timefold.solver.core.config.solver.SolverConfig;
+import ai.timefold.solver.core.config.solver.termination.TerminationConfig;
+import ai.timefold.solver.core.impl.testdata.domain.declarative.fsr.TestdataFSRAssertionEasyScoreCalculator;
+import ai.timefold.solver.core.impl.testdata.domain.declarative.fsr.TestdataFSRConstraintProvider;
+import ai.timefold.solver.core.impl.testdata.domain.declarative.fsr.TestdataFSRRoutePlan;
+import ai.timefold.solver.core.impl.testdata.domain.declarative.fsr.TestdataFSRVehicle;
+import ai.timefold.solver.core.impl.testdata.domain.declarative.fsr.TestdataFSRVisit;
+import ai.timefold.solver.core.preview.api.domain.variable.declarative.ShadowVariableSessionFactory;
+
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+public class FSRShadowVariableTest {
```

Comment:
These tests are complex enough that they deserve comments.
What behavior is being tested by each group of lines, why is the expected result what it is.